### DISCUSSION TITLE
Linking OpenSSL by default breaks Package Control 3.0

### DISCUSSION THREAD

**Discussion by MatthiasPortzel** [Homebrew/discussions#4863] [Everyday usage] [answered]
### Output of `brew config`

<details>
<summary>Details</summary>

```shell
HOMEBREW_VERSION: 4.1.17-9-g141ac15
ORIGIN: https://github.com/Homebrew/brew
HEAD: 141ac157fddd0dbd6c3331c8cc8751412a9acf05
Last commit: 21 hours ago
Core tap JSON: 25 Oct 00:19 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.9ESmYy6Sgl/org.xquartz:0
HOMEBREW_EDITOR: subl
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 12.6.9-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

</details>

### Output of `brew doctor`

<details>
<summary>Details</summary>

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  exa

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/fuse-ext2.pc

Warning: You have uncommitted modifications to Homebrew/brew.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /usr/local/Homebrew && git stash -u && git clean -d -f

Uncommitted files:
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/activesupport-6.1.7.3/MIT-LICENSE
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/addressable-2.8.1/LICENSE.txt
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/connection_pool-2.4.0/LICENSE
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/domain_name-0.5.20190701/LICENSE.txt
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/http-cookie-1.0.5/LICENSE.txt
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/i18n-1.12.0/MIT-LICENSE
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/mechanize-2.8.5/LICENSE.txt
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/nokogiri-1.13.10-x86_64-darwin/LICENSE-DEPENDENCIES.md
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/nokogiri-1.13.10-x86_64-darwin/LICENSE.md
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/public_suffix-5.0.1/LICENSE.txt
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/ruby-macho-3.0.0/LICENSE
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubyntlm-0.6.3/LICENSE
  ?? Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/webrick-1.8.1/LICENSE.txt

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

Warning: Some installed formulae are missing dependencies.
You should `brew install` the missing dependencies:
  brew install openssl@1.1

Run `brew missing` for more details.
```
</details>

### Description of issue

Correct me if I’m wrong, this is the timeline and causation that I’ve put together "posthumously." 

Package Control (Sublime Text’s community-run package manager) relies on system OpenSSL 1.1 binaries in order to work on macOS. If OpenSSL is missing or is upgraded to OpenSSL 3.0, Package Control “disappears.” This is tracked on the Package Control side here: https://github.com/wbond/package_control/issues/1612

About 4 months ago Homebrew migrated from OpenSSL 1.1 to OpenSSL 3 by default, and as part of that they started linking by default (https://github.com/Homebrew/homebrew-core/issues/104784). This increased the number of people who hit Package Control 1612 from just those who manually linked OpenSSL 3, to everyone who installed any Homebrew package which depended on OpenSSL.

Now Package Control has an update in beta which allows it to work with an internal version of OpenSSL 3. And so the solution is for Package Control users to update to that beta until that version of package control is released. I don’t think any action is required on the Homebrew side. I’m just creating this discussion so that the Homebrew team is aware of the speedbumps and fallout of the upgrade, and to inquire whether they knew this was going to break third-party applications (those that were not installed through Homebrew).

Additional notes:
Reproduction is installing Sublime Text, installing Package Control (cmd+shift+p > install package control), installing Homebrew, installing python (or another package that depends on OpenSSL) through Homebrew. Observe that "Package Control: Install Package" is missing from the command palette in Sublime Text.
This doesn't seem to effect ARM-Macbooks. (Although I also don't know if it's been conclusively tested.)

**Comment by SMillerDev** [comment]
> I’m just creating this discussion so that the Homebrew team is aware of the speedbumps and fallout of the upgrade, and to inquire whether they knew this was going to break applications.

I'd imagine that everyone who had the expectation that `/usr/local/lib/libcrypto.dylib` would forever be OpenSSL v1.1, or made assumptions about the general state of people's Homebrew install ran into issues with the update of tools over the years. It's not really something I consider when maintaining packages, there are simply too many unexpected ways people use Homebrew.

**Reply by Bo98** [reply, marked as answer]
Yeah this is definitely upstream's bug. You can rely on `/usr/local/opt/openssl@1.1/lib` to be constant ABI. You absolutely cannot rely on anything `/usr/local/lib` to never change ABI - even from things that are not Homebrew as it's a common install directory for many things.

We will also definitely not be reverting to 1.1 because that's now EOL, deprecated and will be removed entirely soon. The migration to OpenSSL 3 was already as late as could possibly have been given we finished it just weeks before EOL.

Package Control should ideally not blindly run unversioned dylibs. Instead it should search for ABI versions it supports, i.e. `libcrypto.1.1.dylib` rather than the `libcrypto.dylib` symlink. If this is not possible, an alternative could be to set `DYLD_FALLBACK_LIBRARY_PATH=/usr/lib` internally for a more stable approach that avoids `/usr/local/lib`.

**Reply by MatthiasPortzel** [reply]
Thanks for the answers. Since Package Control doesn't know about Homebrew I was surprised that it was being broken by a Homebrew update but it sounds like `/usr/local/lib/` is free game for Homebrew to write to. And yes, there is a patch upstream already that just needs to be deployed (but the maintainer is MIA, of course).
